src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -85,80 +123,115 @@ private MemoryScope() {
-     * This method may only be called in "owner" thread.
-     * The returned instance may be published unsafely to and used in any thread, but methods that explicitly state that
-     * they may only be called in "owner" thread, must strictly be called in single thread that has been selected to be the
-     * "owner" thread.
+     * Duplicates this scope with given new "owner" thread and {@link #close() closes} it.
+     * If this is a root scope, a new root scope is returned; this root scope is closed, but
+     * without executing the cleanup action, which is instead transferred to the duped scope.
+     * If this is a child scope, a new child scope is returned.
+     * This method may only be called in the "owner" thread of this scope unless the
+     * scope is a root scope with no owner thread - i.e. is not checked.
+     * The returned instance may be published unsafely to and used in any thread,
+     * but methods that explicitly state that they may only be called in "owner" thread,
+     * must strictly be called in given new "owner" thread
+     * or else IllegalStateException is thrown.
      *
+     * @param newOwner new owner thread of the returned MemoryScope
      * @return a duplicate of this scope
+     * @throws NullPointerException  if given owner thread is null
      * @throws IllegalStateException if this scope is already closed or if this is
      *                               a root scope and there is/are still active child
-     *                               scope(s).
+     *                               scope(s) or if this method is called outside of
+     *                               owner thread in checked scope
      */
-    abstract MemoryScope dup();
+    abstract MemoryScope dup(Thread newOwner);
+
+    /**
+     * Returns "owner" thread of this scope.
+     *
+     * @return owner thread (or null for unchecked scope)
+     */
+    final Thread ownerThread() {
+        return owner;
+    }
 
     /**
      * This method may be called in any thread.
      *
      * @return {@code true} if this scope is not closed yet.
      */
-    final boolean isAliveThreadSafe() {
+    final boolean isAlive() {
         return !((boolean)CLOSED.getVolatile(this));
     }
 
+    /**
+     * Checks that this scope is still alive and that this method is executed
+     * in the "owner" thread of this scope or this scope is unchecked (not associated
+     * with owner thread).
+     *
+     * @throws IllegalStateException if this scope is already closed or this
+     *                               method is executed outside owning thread
+     *                               in checked scope
+     */
+    @ForceInline
+    final void checkValidState() {
+        if (owner != null) {
+            if (owner != Thread.currentThread()) {
+                throw new IllegalStateException("Attempted access outside owning thread");
+            }
+            checkAliveConfined(this);
+        }
```

Comment:
This changes the behavior we had before in AbstractMemorySegmentImpl::checkValidState. We should still do the liveness check even if there is no owner.

Reply:
> This changes the behavior we had before in AbstractMemorySegmentImpl::checkValidState. We should still do the liveness check even if there is no owner.

True - I suspect Peter would argue that such check doesn't make much sense in a racy context - but still, there's a change in behavior here.

Reply:
IMO It's not necessarily racy, it's just that the API doesn't guarantee that there is no race, that's up to the user to do. e.g. by making sure that the thread doing the access can see an up to date liveness flag by using other synchronization.

Reply:
Ok, then another change follows...